### PR TITLE
Add features and organize code to be more customizable

### PR DESCRIPTION
**Pull request by zhshize**
I think that **Customizable** is the most important property when I am applying this library to many old projects.
So I do many modifications, it will be fantastic!

I merged new updates during development, don't worry about the inconsistency between two repos.

The doc is added in README.md, it is helpful.

I think that the API changes are so significant that it **could be a major version release**.

# What's inside

Fixes:
 - Typos

Enhancements:
 - Type declarations
 - Documentation & PHPDocs
 - Add PHP version requirement
 - Move the interfaces to Contacts namespace (shows that these interfaces are good to implement and replace)
 - Move exception classes to Exceptions namespace
 - Add events, all OIDC and authentication related events will be dispatched
 - Add `Contracts\OIDCAuthenticatable` to mark user object with access token features
 - Split `OIDCAuthServiceProvider` into many tiny functions for customization (extends and small overwritings)
 - Split `CallbackController` into many tiny functions for customization (extends and small overwritings)

Features:
 - Refreshing access token (a function for refresh manually and an option for refresh automatically)
 - Add Facade `OIDC`

Modifications:
 - New OIDCSessionGaurd (doesn't extend Laravel SessionGuard for being concise)
 - All codes about authorization code flow are moved into the class `AuthorizationCodeFlow`
 - All codes about storing token are moved into the class `TokenStore`

# Significant Updates

 - `config/oidc-auth.php` **renamed** to `config/oidc_auth.php`  
 Because of Laravel naming convention
 - `UserFactoryInterface` moved to `Contracts\OIDCAuthenticatableFactory`
 - Middleware `Authenticate` moved to `Middleware\Authenticate`
 - `SessionGuard` **replaced** by `OIDCSessionGaurd `
 - `SessionUserProvider` renamed to `OIDCSessionUserProvider`
 - `InvalidStateException`, `AuthenticationException`, `AuthenticationErrorException` moved to Exceptions namespace
 
 Give me some comment, I will be glad to improve the code.

### CODE REVIEW

**Comment by xdavidwu** [comment]
> I think that the API changes are so significant that it could be a major version release.

Per SemVer, we won't bump major unless we feel easy to gurantee API stability.

> config/oidc-auth.php renamed to config/oidc_auth.php
> Because of Laravel naming convention

Is that a thing in `config/` naming? A quick `find` in my work environment over `vendor/` of a few laravel projects shows a case with `-` (ide-helper.php in barryvdh/laravel-ide-helper, I believe this is popular), and nothing with a `_`.

This is a huge patchset and require time to review. You should probably split the changes into individual PRs next time.

File: composer.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
         }
     ],
     "require": {
+        "php": "^7.4|^8.0",
```
Comment (xdavidwu) [inline composer.json:18]: Do we need to be at least 7.4? The most modern addition in PHP we use, iirc, is null coalescing operator, which is available since 7.0.

Comment (xdavidwu) [reply composer.json:18]: Oh I see you use typed properties (7.4+) in follow-ups

File: src/OIDCSession.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+namespace LaravelOIDCAuth;
+
+use Lcobucci\JWT\Token;
+use OpenIDConnectClient\AccessToken;
+
+class OIDCSession implements OIDCSessionInterface
+{
+    private string $sessionKey = 'oidc-auth.access_token';
+
+    public function user(): ?OIDCUser
```
Comment (xdavidwu) [inline src/OIDCSession.php]: I don't think this one will be useful

File: src/AuthorizationCodeFlow.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+     */
+    public function redirectToAuthorize(): RedirectResponse
+    {
+        $url = $this->provider->getAuthorizationUrl();
+        session()->flash('oidc-auth.state', $this->provider->getState());
+        return response()->redirectTo($url);
+    }
+
+    /**
+     * After the OpenID provider returned an Authentication Response, call this function with Request
```
Comment (xdavidwu) [inline src/AuthorizationCodeFlow.php]: OpenID Connect != OpenID

File: src/OIDCSessionGuard.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                if ($this->OIDCSession->autoRefresh()) {
+                    try {
+                        $token = $this->OIDCSession->accessToken();
+                        $this->OIDCSession->setAccessToken($this->flow->refreshAccessToken($token));
+                        return true;
+                    } catch (\Exception $e) {
+                        return false;
+                    }
+                } else {
+                    $this->OIDCSession->clearAccessToken();
```
Comment (xdavidwu) [inline src/OIDCSessionGuard.php]: side-note: This is a behavoir change that enforce expiration validation futher in already authenticated session lifecycle.

File: config/oidc-auth.php [hunk 5]
```diff
@@ -41,7 +41,7 @@
     |--------------------------------------------------------------------------
     |
     | Factory to get an Illuminate\Contracts\Auth\Authenticatable to use, see
-    | LaravelOIDCAuth\UserFactoryInterface.
+    | LaravelOIDCAuth\Contracts\UserFactoryInterface.
```
Comment (xdavidwu) [inline config/oidc-auth.php]: Not fully renamed here

File: src/OIDCSessionGuard.php [hunk 6]
```diff
@@ -37,6 +37,8 @@ public function user(): ?Authenticatable
                 $this->user = $this->userFactory->authenticatable($this->OIDCSession->accessToken());
             }
             return $this->user;
+        } else {
+            $this->OIDCSession->clearAccessToken();
```
Comment (xdavidwu) [inline src/OIDCSessionGuard.php]: When `tryRefreshToken()` returns false, there is already no token, either cleared by it or not there in the first place.

File: src/OIDCAuthServiceProvider.php [hunk 7]
```diff
@@ -2,41 +2,88 @@
...
 use Illuminate\Support\Facades\Auth;
 use Illuminate\Support\ServiceProvider;
 use LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory;
 
 class OIDCAuthServiceProvider extends ServiceProvider
 {
+    /**
+     * Register OIDC Auth services.
+     *
+     * Please override other functions in this class instead of boot() to customize.
```
Comment (xdavidwu) [inline src/OIDCAuthServiceProvider.php:15]: I don't think it is useful to customize this class.  I don't think any reason doing this, as it just registers stuff and not actually enforce using them.

File: src/CallbackController.php [hunk 8]
```diff
@@ -2,29 +2,100 @@
...
 use OpenIDConnectClient\OpenIDConnectProvider;
 
 class CallbackController extends Controller
 {
     protected AuthorizationCodeFlowInterface $flow;
 
+    /**
+     * @var bool Indicated whether refresh the access token when token expired
+     */
+    protected bool $autoRefresh = false;
```
Comment (xdavidwu) [inline src/CallbackController.php]: This should come from config.

Also I'm not interested in maintaining "customizability" in this class. I don't see valuable reasons doing so, and I am not a fan of making every tiny step a method. The logic is already clear.

File: src/Events/OIDCAuthenticated.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace LaravelOIDCAuth\Events;
+
+use Illuminate\Broadcasting\InteractsWithSockets;
+use Illuminate\Foundation\Events\Dispatchable;
+use Illuminate\Http\Request;
+use Illuminate\Queue\SerializesModels;
+
+/**
+ * Represent the event when OpenID Provider authenticated and sent the authorization code
```
Comment (xdavidwu) [inline src/Events/OIDCAuthenticated.php]: OpenID Connect != OpenID

File: src/Events/OIDCAuthenticated.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+ */
+class OIDCAuthenticated
+{
+    use Dispatchable, InteractsWithSockets, SerializesModels;
+
+    /**
+     * @var Request|null The request returned by OpenID Provider after authorization redirected
```
Comment (xdavidwu) [inline src/Events/OIDCAuthenticated.php]: ditto

File: src/Events/OIDCAuthenticationFailed.php [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+use Illuminate\Http\Request;
+use Illuminate\Queue\SerializesModels;
+use Throwable;
+
+class OIDCAuthenticationFailed
+{
+    use Dispatchable, InteractsWithSockets, SerializesModels;
+
+    /**
+     * @var Request|null The request returned by OpenID Provider after authorization redirected
```
Comment (xdavidwu) [inline src/Events/OIDCAuthenticationFailed.php]: ditto

File: src/Events/OIDCAuthenticationStarted.php [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+{
+    use Dispatchable, InteractsWithSockets, SerializesModels;
+
+    /**
+     * @var string The URL for OIDC authorization redirection
+     */
+    public string $authorizationUrl;
+
+    /**
+     * @var string Opaque value used to maintain state between the request and the callback.
```
Comment (xdavidwu) [inline src/Events/OIDCAuthenticationStarted.php]: Sounds vague. I think if you want to explain, you should just link to related section of OIDC/OAuth2 spec/RFC.

File: src/AccessTokenRefreshable.php [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+ * Add refreshAccessToken function to an OIDCAuthenticatable
+ *
+ * This trait must be added to a class with HasAccessToken trait.
+ * If the default guard is an OIDCSessionGuard, refreshAccessToken() will automatically to call Auth::setUser()
+ *
+ * @see OIDCAuthenticatable
+ * @see HasAccessToken
+ * @see OIDCSessionGuard
+ */
+trait AccessTokenRefreshable
```
Comment (xdavidwu) [inline src/AccessTokenRefreshable.php]: Naming: RefreshesAccessToken?

**Review by xdavidwu — COMMENTED** [review]
First round of review until `d28bc0642d712fad2ebec9ae1d0808a2cc15ac35`

I see you changing your design over and over again within this PR. There are somethings that I agree with, but also some I have my doubt. I don't see convincing reasons in some of your focus on "customizabilty", and I'm afraid some of those changes reduce simplicity and make this code base less maintainable.

File: composer.json (hunk 1 above)
Comment (zhshize) [reply composer.json:18]: oauth2-openid-connect-client requires 7.4+

https://github.com/steverhoades/oauth2-openid-connect-client#requirements

Comment (xdavidwu) [reply composer.json:18]: It is not needed to bring up used library dependency. That is package manager's job.

The real reason for doing this should be:
> you use typed properties (7.4+) in follow-ups

File: src/AccessTokenRefreshable.php (hunk 13 above)
Comment (zhshize) [reply src/AccessTokenRefreshable.php]: 'RefreshesAccessToken' is better, I'll rename it.

File: src/AuthorizationCodeFlow.php (hunk 3 above)
Comment (zhshize) [reply src/AuthorizationCodeFlow.php]: 'OpenID provider' is a term in spec.

https://openid.net/specs/openid-connect-core-1_0.html#Terminology

File: src/Events/OIDCAuthenticationStarted.php (hunk 12 above)
Comment (zhshize) [reply src/Events/OIDCAuthenticationStarted.php]: Modify the comment

File: config/oidc_auth.php [hunk 14]
```diff
@@ -30,25 +30,48 @@
...
+
     /*
     |--------------------------------------------------------------------------
     | Authenticatable Factory
     |--------------------------------------------------------------------------
     |
-    | Factory to get a Illuminate\Contracts\Auth\Authenticatable to use, see
-    | LaravelOIDCAuth\UserFactoryInterface.
-    | For example, you can use a Eloquent model as Authenticatable to store
+    | Factory to get an Illuminate\Contracts\Auth\Authenticatable to use, see
```
Comment (zhshize) [inline config/oidc_auth.php]: OIDCAuthenticatable

**Comment by zhshize** [comment]
To Do:
- List breaking changes
- Fold CallbackController, TokenStore, OIDCAuthServiceProvider
- Remove OIDC Facade
- Other issues in above conversations
- Rearrange commit history
- Rebase from main

**Comment by zhshize** [comment]
### Breaking changes
#### Renamed & moved files
- `config/oidc-auth.php` renamed to `config/oidc_auth.php`
- Exceptions was moved from top level namespace `LaravelOIDCAuth` to `LaravelOIDCAuth\Exceptions`:
  - `AuthenticationErrorException`
  - `AuthenticationException`
  - `InvalidStateException`
- Middleware `LaravelOIDCAuth\Authenticate` moved to `LaravelOIDCAuth\Middleware\Authenticate`
#### Removed files
- Class `OIDCProviderService` was removed
- Class `SessionUserProvider` was removed
#### UserFactory
The UserFactory must implement `LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory`. `UserFactoryInterface` was removed.

**Comment by xdavidwu** [comment]
about addition of composer.lock: I think we don't actually need to use lock file in library development. It will not be respected when used as a library anyway. I am against tracking a lock file as it may hinder potential chances on catching problems on development.

**Comment by xdavidwu** [comment]
> config/oidc-auth.php renamed to config/oidc_auth.php

I'm still against this until this convention is documented in official documents, or there is widespread adoption.

>  LaravelOIDCAuth\OIDCAuthenticatableFactory

afaict it's `LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory`

File: README.md [hunk 15]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
-Use `LaravelOIDCAuth\Authenticate` as auth middleware to redirect directly to OIDC login automatically. Alternatively, you can get login url by yourself and store state string into session storage `oidc-auth.state`, see that middleware.
+Use `LaravelOIDCAuth\Middleware\Authenticate` as auth middleware to redirect directly to OIDC login automatically. Alternatively,
+you can get login url by yourself and store state string into session storage `oidc-auth.state`, see that middleware.
 
 ### `config/auth.php`
 
 Set guard driver to `oidc-auth-session` to remove tokens from session storage on logout.
 
-Set provider driver to `oidc-auth-session` to make `Auth::user()` return a `LaravelOIDCAuth\OIDCUser` authenticable from saved token.
+Set provider driver to `none`, user provider is not suitable.
```
Comment (xdavidwu) [inline README.md]: > is not suitable

is not needed

Edit: `is not used` should be better

> Set provider driver to `none`

Removing that key may be better

File: README.md [hunk 16]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+## Documentation
+
+### Tokens
+
+You can use authenticated user object to get tokens.
+
+```php
+// Get login user, returns \LaravelOIDCAuth\Contracts\OIDCAuthenticatable|null
+Auth::user();
+// Get the saved access token, returns \OpenIDConnectClient\AccessToken|null
```
Comment (xdavidwu) [inline README.md]: Is null possible on `accessToken()` (and also those behind that) with default implementation?

File: README.md [hunk 17]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+// Get the saved refresh token, returns string|null
+Auth::user()->accessToken()->getRefreshToken();
+```
+
+### ID Token
+
+You can get id token claims like this:
+
+```php
+Route::get('/user', function () {
```
Comment (xdavidwu) [inline README.md]: For consistency with previous examples, we should not include route definitions.

File: README.md [hunk 18]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+    return Auth::user()->accessToken()->getIdToken()->getClaims();
+});
+```
+
+### Refresh the access token
+
+If you are using `oidc-auth-session` guard with `OIDCUser`, you can easily refresh the access token:
```
Comment (xdavidwu) [inline README.md]: > easily refresh the access token:

refresh access token like following:

File: README.md [hunk 19]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+
+```php
+use LaravelOIDCAuth\Facades\OIDC;
```
Comment (xdavidwu) [inline README.md]: facade not actually used

File: README.md [hunk 20]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+// ...
+Auth::user()->refreshAccessToken();
+```
+
+Typically, you don't need to refresh the access token manually, the following doc shows an automatic way to do this.
```
Comment (xdavidwu) [inline README.md]: > Typically, you don't need to refresh the access token manually, the following doc shows an automatic way to do this.

You can also turn the `Auto refresh`  feature on instead of refreshing manually.

We should also mention that whether token is actually refreshed by that function if it is not expired. If it is always refreshed, we should document how to check expiration against existing token.

Or even better, `Auto refresh` section should come before this one.

File: README.md [hunk 21]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+
+#### Auto refresh
+
+Sometimes, you want the `oidc-auth-session` to refresh the access token automatically every request, please edit
```
Comment (xdavidwu) [inline README.md]: If you need to automatically validate expiration and refresh tokens if possible, enable `auto_refresh` in config.

(No code example required as it is trivial.)

File: README.md [hunk 22]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+`config/oidc_auth.php`:
+
+```php
+    // ...
+    'auto_refresh' => true,
+    // ...
+```
+
+Next time the user logged in, the `CallbackController` will set `rememberMe` option from
```
Comment (xdavidwu) [inline README.md]: Is this rememberMe mechanism related to refreshing token? Why is it tied?

File: README.md [hunk 23]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+`config('oidc_auth.auto_refresh')` to the guard's `login` function.
+
+### User Model & User Factory
```
Comment (xdavidwu) [inline README.md]: For consistency for cases, `User model & user factory`?

File: README.md [hunk 24]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+
+If you need an authenticated user to be an eloquent model with OIDC features, your user model must implement
```
Comment (xdavidwu) [inline README.md]: If a custom user implementation is desired (for example, as an Eloquent model), it should implement ...

File: README.md [hunk 25]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+use Illuminate\Notifications\Notifiable;
+use Illuminate\Contracts\Auth\MustVerifyEmail;
+use LaravelOIDCAuth\Contracts\OIDCAuthenticatable;
+use LaravelOIDCAuth\HasAccessToken;
+use LaravelOIDCAuth\RefreshesAccessToken;
+use Illuminate\Foundation\Auth\User as Authenticatable;use OpenIDConnectClient\AccessToken;
+
+class User extends Authenticatable implements OIDCAuthenticatable
+{
+    use Notifiable, HasAccessToken, RefreshesAccessToken;
```
Comment (xdavidwu) [inline README.md]: Why is Notifiable required?

Also, I don't think code example is required here. Mentioning traits that may help implementation is enough.

File: README.md [hunk 26]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+
+use LaravelOIDCAuth\Contracts\OIDCAuthenticatable;
+use LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory;
+use OpenIDConnectClient\AccessToken;
+
+class UserFactory implements OIDCAuthenticatableFactory
+{
+    public function authenticatable(AccessToken $token): OIDCAuthenticatable
+    {
+        $user = User::firstOrCreate(/* ... */);
```
Comment (xdavidwu) [inline README.md]: Again, no code example should be required. This also assumes an Eloquent model, which may not be the only usecase.

File: README.md [hunk 27]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+        return $user;
+    }
+}
+```
+
+### Customization
+
+#### Handle OIDC callback
+
+When the OIDC Provider authenticated the user, the browser will be redirected to the callback route on your own web
```
Comment (xdavidwu) [inline README.md]: When user authenticates via OIDC,

Edit: When user is authenticated by OIDC Provider, the browser will be redirected to callback route...

Comment (xdavidwu) [reply README.md]: (I don't really think you should explain callback here. Consumer should already know it if they need to handle that manually)

File: README.md [hunk 28]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+application. If you want to handle the callback request manually, you can disable the registration of default callback
+route in `config/oidc_auth.php`:
+
+```php
+    'register_default_callback_route' => false,  // default value is true, set to false
```
Comment (xdavidwu) [inline README.md]: Pointing out defaults here is redundant (you are already pointing out this need to be set manually.)

File: README.md [hunk 29]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+```
+
+Next, you can create your own route & controller to handle OIDC callback, and the route must be named as
+`oidc-auth.callback`.
+
+Here is an example of your callback controller:
+
+```php
+namespace App\Http\Controllers;
```
Comment (xdavidwu) [inline README.md]: Trivial, no code example needed.

File: README.md [hunk 30]
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
...
+    public function callback(Request $request)
+    {
+        // Do something
+        return parent::callback($request);
+    }
+    // You can overwrite other method from parent class LaravelOIDCAuth\CallbackController
+}
+```
+
+Add a route in `route/web.php`:
```
Comment (xdavidwu) [inline README.md]: Also trivial and unnecessarily assumes consumer's route organization.

**Review by xdavidwu — COMMENTED** [review]
A few comments for documentation only

**Comment by zhshize** [comment]
> > config/oidc-auth.php renamed to config/oidc_auth.php
> 
> I'm still against this until this convention is documented in official documents, or there is widespread adoption.
> 

There is an example: https://github.com/laravel/cashier-mollie/tree/develop/config
and a non-official guide: https://github.com/alexeymezenin/laravel-best-practices#follow-laravel-naming-conventions

I like `config('oidc_auth.auto_refresh')` better than `config('oidc-auth.auto_refresh')`.

> > LaravelOIDCAuth\OIDCAuthenticatableFactory
> 
> afaict it's `LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory`

`LaravelOIDCAuth\OIDCAuthenticatableFactory` -> `LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory` updated.

File: README.md (hunk 15 above)
Comment (zhshize) [reply README.md]: I agree.

File: README.md (hunk 17 above)
Comment (zhshize) [reply README.md]: Okay.

**Comment by zhshize** [comment]
> about addition of composer.lock: I think we don't actually need to use lock file in library development. It will not be respected when used as a library anyway. I am against tracking a lock file as it may hinder potential chances on catching problems on development.

I'll remove that commit.

File: README.md (hunk 18 above)
Comment (zhshize) [reply README.md]: Okay.

**Comment by xdavidwu** [comment]
> There is an example: https://github.com/laravel/cashier-mollie/tree/develop/config

While that is not a famous package, it seems like that is the only example with multiple words under laravel namespace (although it is moving out). If we are going for underscores, we should also rename the route definition file.

File: README.md (hunk 19 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 23 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 22 above)
Comment (zhshize) [reply README.md]: Yes, it is.

If the user log in with 'remember me', guard will automatically refresh the access token to keep the user logged in when access token is expired.  Actually, the 'auto_refresh' option is treated as the default value of the rememberMe argument when controller log the user in(See: [loginUser](https://github.com/zhshize/laravel-oidc-auth/blob/53bfe0bfbfc47104e9a7ef6e1c6618b79fb2991f/src/CallbackController.php#L91)).  This mechanism is a simulation of remember me feature when we integrate OIDC to Laravel authentication.

Do you think the mechanism should be described in the doc or this mechanism could be confusing?

File: README.md (hunk 20 above)
Comment (zhshize) [reply README.md]: I rearranged the doc.

File: README.md (hunk 21 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 24 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 25 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 26 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 27 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 28 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 29 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 30 above)
Comment (zhshize) [reply README.md]: Okay.

File: README.md (hunk 22 above)
Comment (xdavidwu) [reply README.md]: Well I see you just removed that in newer commit.

If it is implementation detail, it should not be doc'ed at README. We do not want to gurantee this level of compatibility.

IIRC, "remember me" in Laravel document is some sort of long-lived cookie stuff, which is unrelated to this token refreshing behavior. Better use a more straight-forward naming in code instead of mimicking something that we don't actually implement.

**Comment by xdavidwu** [comment]
Mark resolved threads as resolved (thread status, not by commenting) and I will do another review round.

File: README.md (hunk 22 above)
Comment (zhshize) [reply README.md]: I renamed the parameter remember to auto_refresh in the newest commit.

File: src/OIDCSession.php (hunk 2 above)
Comment (zhshize) [reply src/OIDCSession.php]: Removed

File: config/oidc-auth.php (hunk 5 above)
Comment (zhshize) [reply config/oidc-auth.php]: fixed in newer commit

File: src/Events/OIDCAuthenticated.php (hunk 10 above)
Comment (zhshize) [reply src/Events/OIDCAuthenticated.php]: fixed in newer commit

File: src/Events/OIDCAuthenticated.php (hunk 9 above)
Comment (zhshize) [reply src/Events/OIDCAuthenticated.php]: fixed in newer commit

File: src/Events/OIDCAuthenticationFailed.php (hunk 11 above)
Comment (zhshize) [reply src/Events/OIDCAuthenticationFailed.php]: fixed in newer commit

File: src/CallbackController.php (hunk 8 above)
Comment (zhshize) [reply src/CallbackController.php]: Refactored in newer commit

File: src/OIDCAuthServiceProvider.php (hunk 7 above)
Comment (zhshize) [reply src/OIDCAuthServiceProvider.php:15]: Refactored in newer commit

File: src/AuthorizationCodeFlow.php (hunk 3 above)
Comment (zhshize) [reply src/AuthorizationCodeFlow.php]: Change to 'OpenID Connect provider' in newer commit

File: src/OIDCSessionGuard.php (hunk 4 above)
Comment (zhshize) [reply src/OIDCSessionGuard.php]: Downstream user won't need to validate the expiration, I think that it is a useful change.

File: src/Events/OIDCAuthenticationStarted.php (hunk 12 above)
Comment (zhshize) [reply src/Events/OIDCAuthenticationStarted.php]: fixed in newer commit

File: config/oidc_auth.php (hunk 14 above)
Comment (zhshize) [reply config/oidc_auth.php]: fixed in newer commit

File: src/OIDCSessionGuard.php (hunk 4 above)
Comment (xdavidwu) [reply src/OIDCSessionGuard.php]: I'm not against it. I am just noting it for further release planning.

Please keep this kind of side-noting conversation open.

**Comment by xdavidwu** [comment]
this needs a rebase

**Comment by xdavidwu** [comment]
Closing due to inactivity and lack of consensus.

To whom it may concern, discuss with maintainers before implementing such a big patchset.
